snapshot/git/gitdb/bundlestore.go

```diff
@@ -0,0 +1,200 @@
+package gitdb
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+
+	snap "github.com/scootdev/scoot/snapshot"
+	"github.com/scootdev/scoot/snapshot/bundlestore"
+)
+
+// BundlestoreConfig defines how to talk to Bundlestore
+type BundlestoreConfig struct {
+	Store bundlestore.Store
+}
+
+type bundlestoreBackend struct {
+	cfg *BundlestoreConfig
+}
+
+const bundlestoreIDText = "bs"
+
+// "bs-gc-<bundle>-<stream>-<sha>"
+
+func (b *bundlestoreBackend) parseID(id snap.ID, kind snapshotKind, extraParts []string) (snapshot, error) {
+	if len(extraParts) != 3 {
+		return nil, fmt.Errorf("cannot parse snapshot ID: expected 5 parts in bundlestore id: %s", id)
+	}
+	bundleName, streamName, sha := extraParts[0], extraParts[1], extraParts[2]
+
+	if err := validSha(sha); err != nil {
+		return nil, err
+	}
+
+	return &bundlestoreSnapshot{kind: kind, sha: sha, bundleName: bundleName, streamName: streamName}, nil
+}
+
+func (b *bundlestoreBackend) upload(s snapshot, db *DB) (snapshot, error) {
+	// We only have to upload a localSnapshot
+	switch s := s.(type) {
+	case *tagsSnapshot:
+		return s, nil
+	case *streamSnapshot:
+		return s, nil
+	case *bundlestoreSnapshot:
+		return s, nil
+	case *localSnapshot:
+		return b.uploadLocalSnapshot(s, db)
+	default:
+		return nil, fmt.Errorf("cannot upload %v: unknown type %T", s, s)
+	}
+}
+
+// git bundle create takes a rev list; it requires that it include a postiive ref
+// so we can't just do:
+// git bundle create 545c88d71d40a49ebdfb1d268c724110793330d2..3060a3a519888957e13df75ffd09ea50f97dd03b
+// instead, we have to write a temporary ref
+const bundlestoreTempRef = "reserved_scoot/bundlestore/__temp_for_writing"
+
+func (b *bundlestoreBackend) uploadLocalSnapshot(s *localSnapshot, db *DB) (sn snapshot, err error) {
+	// the sha of the commit we're going to use as the ref
+	commitSha := s.sha
+
+	// the revList to create the bundle
+	// unless we can find a better merge base, we'll just include the commit
+	revList := bundlestoreTempRef
+
+	// the name of the stream that this bundle requires
+	streamName := ""
+
+	switch s.kind {
+	case kindGitCommitSnapshot:
+		if db.stream.cfg != nil && db.stream.cfg.RefSpec != "" {
+			streamHead, err := db.dataRepo.Run("rev-parse", db.stream.cfg.RefSpec)
+			if err != nil {
+				return nil, err
+			}
+
+			mergeBase, err := db.dataRepo.RunSha("merge-base", streamHead, commitSha)
+
+			// if err != nil, it just means we don't have a merge-base
```

Comment:
Elaborate on what it means to not have a merge-base. We have a desired s.sha (*are we assuming sha is validated beforehand / should we?*) and if it doesn't share an ancestor with streamHead, we create a bundle with only s.sha? When would this happen?

If we have a bunch of exclusively local commits and there's no ancestor in common with streamHead, isn't the generated bundle useless?